### PR TITLE
Add l3keys2e code as ltkeys

### PR DESCRIPTION
**Pull request by josephwright**
There are a few internal changes, but the main obvious
one is that the family name is made optional:
this will likely help 'detect' that code has been updated.

**READ ME FIRST**: Please understand that in most cases we will not be able to merge a pull request because there are a lot of internal activities needed when updating the LaTeX2e sources. If you have a code suggestion please discuss it with the team first.

*Pull requests in this repository are intended for LaTeX Team members only.*

# Internal housekeeping

## Status of pull request

-  Feedback wanted 
- Under development
- [X] Ready to merge

## Checklist of required changes before merge will be approved
- [X] Test file(s) added
- [X] Version and date string updated in changed source files
- [X] Relevant `\changes` entries in source included
- [X] Relevant `changes.txt` updated
- [X] `ltnewsX.tex` (and/or `latexchanges.tex`) updated

### CODE REVIEW

**Comment by josephwright** [comment]
At the moment I've gone with keeping `Keys` as the name, but that does mean I have to worry about the difference between the 'old' and 'new' `\ProcessKeysOptions'. We might want to just use `Key`.

**Comment by josephwright** [comment]
I'm not entirely sure if we want `DeclareKeysOptions` to allow 'underlying' `l3keys` properties, or to be entirely self-contained. The latter means reproducing some of the parsing, although not the rest of the approach. (For example, I could just regex the key name to swap out `.if` for `.legacy_if:n`, as it has to be at the end of the name. Doubtless this could be done faster without a regex!)

**Comment by josephwright** [comment]
Another open question: do we want the `'PackageOptions`/`Options` split? One might argue that global options should be left to classes, so just have the `PackageOptions` version.

File: base/testfiles/github-0710.lvt [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+\OMIT
+\begin{document}
+\TIMO
+
+\SetKeys[mypkg]{general-option-C = 999,
+  preamble-option-D = 999} % 9 
+\mypkgtest
+
+\END
```
Comment (FrankMittelbach) [inline base/testfiles/github-0710.lvt:50]: missing newline at end of file

File: base/ltkeys.dtx [hunk 2]
```diff
@@ -0,0 +1,454 @@
...
+%    \end{macrocode}
+%
+%    \begin{macrocode}
+%<*2ekernel>
+%    \end{macrocode}
+%
+%    \begin{macrocode}
+\ExplSyntaxOn
+%    \end{macrocode}
+%
```
Comment (FrankMittelbach) [inline base/ltkeys.dtx:151]: this should of course use the declaremodule method for rollback (but I understand why that is not yet there :-)

**Review by FrankMittelbach — COMMENTED** [review]
looks good to me ...

made a few comments (some got outdated while reading further, but I may have forgotten to take out all of the obsolete ones) - so read it with that in mind (that wrote them in order of reading.

File: base/format.ins [hunk 3]
```diff
@@ -203,6 +203,7 @@ the system are in the document `cfgguide.tex'.
           \from{ltbibl.dtx}{2ekernel}
           \from{ltpage.dtx}{2ekernel}
          \from{ltclass.dtx}{2ekernel,tracerollback}
+          \from{ltkeys.dtx}{2ekernel}           % L3 layer module
           \from{ltfilehook.dtx}{2ekernel}       % L3 layer module
```
[the remark this answers is not in the document]
Comment (josephwright) [reply base/format.ins:207]: Comes after `ltclasses` in both?

[the remark this answers is not in the document]
Comment (FrankMittelbach) [reply base/format.ins:207]: must have had tomatos on my eyes ... I could have sworn it was different :-)

File: base/ltkeys.dtx [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+\clist_new:N \l_@@_options_clist
+%    \end{macrocode}
+% \end{macro}
+%
+% \begin{variable}{\l_@@_options_loading_bool}
+%   Used to indicate we are in the loading phase: controls the outcome
+%   of warnings.
+%    \begin{macrocode}
+\bool_new:N \l_@@_options_loading_bool
+%    \end{macrocode}
```
[the remark this answers is not in the document]
Comment (josephwright) [reply base/ltkeys.dtx:169]: This is used specifically to indicate that keys are being set by `\ProcessKeyOptions`, not for anything else. So there are only two states.

File: base/ltkeys.dtx (hunk 2 above)
Comment (josephwright) [reply base/ltkeys.dtx:151]: A policy thing I guess :) I was imagining that as this was strictly opt-in, it didn't need to be rolled back.

File: base/ltkeys.dtx (hunk 4 above)
[the remark this answers is not in the document]
Comment (FrankMittelbach) [reply base/ltkeys.dtx:169]: sorry, one of the leftovers, forgot to take that out once I realzed that I'm probably wrong.

File: base/ltkeys.dtx (hunk 2 above)
Comment (FrankMittelbach) [reply base/ltkeys.dtx:151]: Even then it should have a rollback and thiss way also a roll forward. For example, when Overleaf is a year behind. Stating that this here is a new module makes this fairly painless.

Comment (FrankMittelbach) [reply base/ltkeys.dtx:151]: I maintain that making this a rollback module is essential and in line with our policy to support rolling forward, e.g., on overleaf. So I think we should do this. I can have a go at it if you don't want to.

**Review by FrankMittelbach — APPROVED** [review]
Has my approval, but I think we should do the rollback setup. However, that could be done a little later and we push it out as is into the upcoming dev release, so that we may get some useful feedback earlier.

**Pull request merged**